`includeHTML()` warns if complete document detected

gadenbuie opened the pull request:
Fixes #124 by emitting a warning when `includeHTML()` is passed a path to a file containing a complete HTML document.

This PR originally contained an implementation of `includeHTMLDocument()`. I'll open a new issue to track progress on that front, but I didn't want to hold up this change on working out the details we'd need to address there.

## Example

`includeHTML()` now warns with a recommendation to use `includeHTMLDocument()`

``` r
# Create a temporary complete HTML file
tmp_html <- withr::local_tempfile(fileext = ".html")
save_html(p("Test"), tmp_html)

includeHTML(tmp_html) |> print(browse = FALSE)
#> Warning message:
#> `includeHTML()` was provided a `path` that appears to be a complete HTML document.
#> ✖ Path: /var/folders/yl/0xhq7c7n3z7_9hp1356kctbr0000gp/T//Rtmp9wJCqR/filee0569372d04.html
#> ℹ Use `tags$iframe()` to include an HTML document. You can either ensure `path`
#>   is accessible in your app or document (see e.g. `shiny::addResourcePath()`)
#>   and pass the relative path to the `src` argument. Or you can read the 
#>   contents of `path` and pass the contents to `srcdoc`. 
#> <!DOCTYPE html>
#> <html lang="en">
#> <head>
#> <meta charset="utf-8"/>
#> <style>body{background-color:white;}</style>
#> 
#> 
#> </head>
#> <body>
#> <p>Test</p>
#> </body>
#> </html>
```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/rstudio/htmltools?pullRequest=382) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/rstudio/htmltools?pullRequest=382) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/rstudio/htmltools?pullRequest=382) it.</sub>

cpsievert commented on R/tags.R:
```diff
@@ -1734,15 +1734,81 @@ knit_print.html_dependency <- knit_print.shiny.tag
+    iframe_args$src <- as.character(path)
+  } else {
+    lines <- readLines(path, warn=FALSE, encoding='UTF-8')
+    iframe_args$srcdoc <- paste8(lines, collapse='\n')
```
Let's make sure this works well with `save_html()` on a `{htmlwidgets}`

gadenbuie commented on R/tags.R:
```diff
@@ -1734,15 +1734,81 @@ knit_print.html_dependency <- knit_print.shiny.tag
+
+  iframe_args <- utils::modifyList(iframe_args_default, dots_list(...))
+
+  if (inherits(path, "AsIs") || grepl("^http?s://", path)) {
```
```suggestion
  if (inherits(path, "AsIs") || grepl("^https?://", path)) {
```

cpsievert commented on R/tags.R:
```diff
@@ -1734,15 +1734,81 @@ knit_print.html_dependency <- knit_print.shiny.tag
+    iframe_args$srcdoc <- paste8(lines, collapse='\n')
+  }
+
+  return(tags$iframe(class = "html-fill-item", !!!iframe_args))
```
```suggestion
  bindFillRole(tags$iframe(!!!iframe_args), fill = fill)
```

cpsievert commented on R/tags.R:
```diff
@@ -1734,15 +1734,81 @@ knit_print.html_dependency <- knit_print.shiny.tag
+#'
+#' @rdname include
+#' @export
+includeHTMLDocument <- function(path, ...) {
```
```suggestion
includeHTMLDocument <- function(path, ..., height = NULL, fill = is.null(height)) {
```

schloerke commented on R/tags.R:
```diff
@@ -1734,15 +1734,81 @@ knit_print.html_dependency <- knit_print.shiny.tag
+  TRUE
+}
+
+#' @note `includeHTMLDocument()` embeds the contents of the file in an
```
What is the difference between `@note` and `@describeIn`? Could we use `@describeIn include DESC` here?

schloerke replied:
I guess this would have to be applied to the whole file.

No need to change it, still curious about the difference though.

gadenbuie commented:
Thanks for the input, I'm going to move this to draft while I consider next steps.

gadenbuie commented:
@cpsievert @schloerke I updated this PR to reduce the scope and address the immediate problem posed in #124. With this change, `includeHTML()` will issue a warning if it detects that the user is attempting to include a complete HTML document. 

I've opened a new issue #383 to track thoughts and progress toward a new `includeHTMLDocument()` function, which uses the implementation I've removed from this PR.

wch commented:
Looks good -- just needs a NEWS entry.
